src/Icicle/Common/Exp/Compounds.hs

```diff
@@ -216,21 +217,41 @@ substMaybe name payload into
 
 -- | Substitute an expression in,
 -- using fresh names to avoid capture
-subst :: (Hashable n, Eq n)
+subst1 :: (Hashable n, Eq n)
       => a
       -> Name n
       -> Exp a n p
       -> Exp a n p
       -> Fresh n (Exp a n p)
-subst a_fresh name payload into
+subst1 a_fresh name payload into
+ = subst a_fresh (Map.singleton name payload) into
+
+-- | Substitute an expression in,
+-- using fresh names to avoid capture
+subst :: (Hashable n, Eq n)
+      => a
+      -> Map.Map (Name n) (Exp a n p)
+      -> Exp a n p
+      -> Fresh n (Exp a n p)
+subst a_fresh env into
+ | Map.null env
+ = return into
+ | otherwise
+ = subst' a_fresh env (Set.unions $ Map.elems $ Map.map freevars env) into
+
+subst'  :: (Hashable n, Eq n)
+        => a
+        -> Map.Map (Name n) (Exp a n p)
+        -> Set.Set (Name n)
+        -> Exp a n p
+        -> Fresh n (Exp a n p)
+subst' a_fresh env frees into
```

Comment:
ah, okay. I am not sure - I tried just switching to `Data.Map.Strict` and no real differences